modules/core/src/main/scala/parser.scala

```diff
@@ -312,71 +316,145 @@ object GraphQLParser {
+
+  def token0[A](p: Parser0[A]): Parser0[A] =
+    p <* skipWhitespace <* many(comment)
 
   /**
    * Consumes `left` and `right`, including the trailing and preceding whitespace,
    * respectively, and returns the value of `p`.
    */
-  private def _bracket[A,B,C](left: Parser[B], p: => Parser[A], right: => Parser[C]): Parser[A] =
-    token(left) ~> token(p) <~ token(right)
+  private def _bracket[A,B,C](left: Parser[B], p: => Parser0[A], right: => Parser[C]): Parser[A] =
+    token(left) *> token0(p) <* token(right)
 
   /** Turns a parser into one that consumes surrounding parentheses `()` */
-  def parens[A](p: => Parser[A]): Parser[A] =
-    _bracket(char('('), p, char(')')).named(s"parens(${p.toString})")
+  def parens[A](p: => Parser0[A]): Parser[A] =
+    _bracket(char('('), p, char(')')).withContext(s"parens(${p.toString})")
 
   /** Turns a parser into one that consumes surrounding curly braces `{}` */
-  def braces[A](p: => Parser[A]): Parser[A] = {
+  def braces[A](p: => Parser0[A]): Parser[A] = {
     lazy val q = p
-    _bracket(char('{'), q, char('}')).named(s"braces(${q.toString})")
+    _bracket(char('{'), q, char('}')).withContext(s"braces(${q.toString})")
   }
 
   /** Turns a parser into one that consumes surrounding square brackets `[]` */
-  def squareBrackets[A](p: => Parser[A]): Parser[A] = {
+  def squareBrackets[A](p: => Parser0[A]): Parser[A] = {
     lazy val q = p
-    _bracket(char('['), q, char(']')).named(s"squareBrackets(${q.toString})")
+    _bracket(char('['), q, char(']')).withContext(s"squareBrackets(${q.toString})")
+  }
+}
+
+object Literals {
+
+  lazy val stringLiteral: Parser[String] = {
+
+    lazy val stringCharacter: Parser[String] = (
+      (peek(not(charIn('"', '\\') | lineTerminator)).with1 *> sourceCharacter) |
+        (string("\\u") ~ escapedUnicode).string |
+        (char('\\') ~ escapedCharacter).string
+    )
+
+    lazy val blockStringCharacter = string("TODO").string
+
+    //TODO should this be a Parser[Char] which converts to unicode? I guess so?
+    //See Atto for reference
+    lazy val escapedUnicode: Parser[String] = (digit | hexdig).repExactlyAs[String](4)
+
+    lazy val escapedCharacter = charIn('"', '\\', '/', 'b', 'f', 'n', 'r', 't')
+
+    lazy val sourceCharacter: Parser[String] = (charIn(0x0009.toChar, 0x000A.toChar, 0x000D.toChar) | charIn(0x0020.toChar to 0xFFFF.toChar)).string
+
+    lazy val lineTerminator = (lf | cr | crlf).string
+
+    stringCharacter.rep0.string.with1.surroundedBy(char('"')) | (blockStringCharacter.rep0.string.with1.surroundedBy(string("\"\"\"")))
+
+  }
+
+  lazy val intLiteral: Parser[Int] =
+    bigInt.flatMap {
+      case v if v.isValidInt => pure(v.toInt)
+      case v => failWith(s"$v is larger than max int")
+    }
+
+  lazy val booleanLiteral: Parser[Boolean] = string("true").as(true) | string("false").as(false)
+
+  lazy val floatLiteral: Parser[Float] = {
+
+    lazy val bigDecimal: Parser[BigDecimal] = for {
```

Comment:
this use of flatMap here is going to be considerably slower for float heavy code, I think.

You could write your parser with only 1 flatMap easily:
```scala
(intLiteral, (char('.') *> digit.rep.string).?, ((char('e') | char('E')) *> intLiteral.string).?)
  .tupled
  .flatMap {
        case (a, Some(b), None) => pure(BigDecimal(s"$a.$b"))
        case (a, None, Some(c)) => pure(BigDecimal(s"${a}E$c"))
        case (a, Some(b), Some(c)) => pure(BigDecimal(s"$a.${b}E$c"))
        case (a, None, None) => failWith(s"$a is not a valid float - must have at least one of a fractional or exponent part") 
  }
```
that said, I think if you compare to this:

https://github.com/typelevel/cats-parse/blob/85f3c7f3ee67f9fe96c7754bd06e81a00d533da1/core/shared/src/main/scala/cats/parse/Numbers.scala#L65

you can see a way to write this with no flatMap.